File: python/DataStructures/LinkedList.py
"""
LinkedList.py

Singly-Linked List implementation, iterable.

"""
from timeit import Timer

class Node:
	def __init__(self, data, next=None):
		self.data = data
		self.next = next

class LinkedList:
	
	def __init__(self):
		'''
			Initialize with head = None
		'''
		self.head = None
	
	def __iter__(self):
		current = self.head
		while current is not None:
			yield current.data
			current = current.next
	
	def insert_head(self, data):
		'''
			Insert at head
			O(1)
		'''
		if self.head is None:
			self.head = Node(data, None)
		else:
			oldhead = self.head
			self.head = Node(data, oldhead)

	def delete_head(self):
		'''
			Delete at head
			O(1)
		'''
		if self.head is None:
			return None
		else:
			oldhead = self.head
			self.head = self.head.next
			return oldhead

	def insert_tail(self, data):
		'''
			Insert at tail
			O(n)
		'''
		if self.is_empty():
			self.insert_head(data)
		else:
			tmp = self.head
			while tmp.next is not None:
				tmp = tmp.next
			tmp.next = Node(data, None)
	
	def delete_tail(self):
		'''
			Delete at tail
			O(n)
		'''
		last = self.head
		secondlast = None
		while tmp.next is not None:
			secondlast = last
			last = last.next

		if secondlast is not None:
			secondlast.next = None
		else:
			self.head = None

		return last

	def is_empty(self):
		return self.head == None

	def __str__(self):
		'''
		Prints the current list in the form of a Python list            
		'''
		current = self.head
		toPrint = []
		while current != None:
			toPrint.append(current.data)
			current = current.next
		return str(toPrint) 


def test():
	LL = LinkedList()
	for i in range(0,20):
		LL.insert_head(i)

if __name__ == "__main__":
    # LL = LinkedList()
    # for i in range(0,20):
    	# LL.insert_head(i)
    # print(LL)

    t = Timer("test()", "from __main__ import test")
    print(t.timeit())

    #Test iterable
    # for item in LL:
    	# print(item)

File: java/src/DataStructures/LinkedList.java
package DataStructures;
import java.util.Iterator;

/**
 * 
 * LinkedList Implementation with O(1) insertion and deletion from front (optimized for LIFO Stack)
 *
 * @param <Type>
 */
public class LinkedList<Type> implements Iterable{
	
	private Node<Type> head;
	private int size = 0;
	
	private static class Node<Type>
	{
		private Type data;
		private Node next;
		public Node(Type data, Node<Type> next)
		{
			this.data = data;
			this.next = next;
		}
	}
	
	public LinkedList()
	{
		this.head = null;
	}
	
	/*
	 * Iterable implementation
	 * @see java.lang.Iterable#iterator()
	 */
	public Iterator iterator() {
		return new LinkedListIterator();
	}
	
	private class LinkedListIterator implements Iterator<Type>
	{
		private Node<Type> current = head;
		
		public boolean hasNext() {
			return current != null;
		}
		
		public Type next() {
			Type item = current.data;
			current = current.next;
			return item;
		}
		
	}
	
	public void insertHead(Type item)
	{
		/*
		 * Add to the front of the list 
		 * O(1)
		 */
		Node oldhead = this.head;
		this.head = new Node(item, oldhead);
		size ++;
	}
	
	public Node deleteHead(Type item)
	{
		/*
		 * Remove from front of the list
		 * O(1)
		 */
		if(head==null) throw new IndexOutOfBoundsException("List is empty"); 
		Node first = this.head;
		this.head = this.head.next;
		size--;
		return first;
	}
	
	public void insertTail(Type item)
	{
		/*
		 * Add to the end of the list. Have to traverse to the end starting from head
		 * O(n)
		 */
		if(head==null){
			insertHead(item);
			return;
		}
		
		Node tmp = head;
		while(tmp.next != null) tmp = tmp.next;
		tmp.next = new Node(item, null);
		size++;
	}
	
	public Node deleteTail(Type item)
	{
		/*
		 * Remove from the end of the list
		 * O(n)
		 */
		if(head==null) throw new IndexOutOfBoundsException("List is empty");
		
		Node tmp = head;
		Node prev = null;
		while(tmp.next != null) 
		{
			prev = tmp;
			tmp = tmp.next;
		}
		head = prev;
		size--;
		return tmp;
	}
	
	public int size()
	{
		return this.size;
	}
	
	public static void main(String[] args)
	{
		LinkedList<Integer> list = new LinkedList<Integer>();
		// populate list
		int i=0;
		while (i < 20)
		{
			list.insertTail(i);
			i++;
		}
		
		Iterator<String> iter = list.iterator();
		while(iter.hasNext())
		{
			System.out.println(String.valueOf(iter.next()));
		}
	}
	
}



